### PR TITLE
Fundamental cycle basis implementations

### LINKED ISSUE #124: Paton cycle detection with pseudographs
`org.jgrapht.alg.cycle.PatonCycleBase` takes an `UndirectedGraph` as an argument and thus doesn't prohibit the use of a `PseudoGraph`. 

However I am encountering a few issues while using it with `PseudoGraph`s. See this commit in my fork: https://github.com/josemduarte/jgrapht/commit/5af6de85155c816bda73f18201dc256b7d2096ab
- **Issue 1**: if given a graph with 2 vertices and 2 edges between them, `PatonCycleBase` finds 1 cycle of size 3 (where the first vertex is repeated twice) when it should be 1 cycle of size 2. 
- **Issue 2/3**: if given a graph with a 4 cycle (a square) and a duplicate edge in one of the sides then `PatonCycleBase.findCycleBase()` goes in an infinite loop, but only sometimes depending on the position of the double edge (see testPseudographIssue2 for a position that works, testPseudographIssue3 for a position that goes to the infinite loop).

Should Paton cycle detection work in PseudoGraphs? if not a check should be added to reject PseudoGraphs with an `IllegalArgumentException`. If PseudoGraphs are valid input then most likely the above issues are bugs.

### PR DESCRIPTION
**Pull request by d-michail**
Improved cycle basis support. Revised cycles to be edge sets and not vertex sets.

 - Added interface for undirected cycle basis algorithms
 - Wrote QueueBFS and StackBFS fundamental cycle basis algorithms
 - Wrote tests for QueueBFS and StackBFS
 - Added hasSelfLoops() and hasMultipleEdges() in GraphTests
 - Made PatonCycleBase conform to the new interfaces
 - Resolves #124 
 - Deprecated UndirectedCycleBase interface
 - Improved package level documentation

Update: 
 - Note that I on purpose did not use a `GraphPath` here, as cycles are vectors in GF2 when being in the context of undirected cycle bases. This means that a cycle is just an edge set where each vertex has even degree.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/QueueBFSFundamentalCycleBasis.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ */
+public class QueueBFSFundamentalCycleBasis<V, E>
+    extends AbstractFundamentalCycleBasis<V, E>
+{
+
+    /**
+     * Compute a spanning forest of the graph using a straightforward BFS implementation.
+     * 
+     * The representation assumes that the map contains the predecessor edge of each vertex in the
+     * forest. The predecessor edge is the forest edge that was used to discover the vertex. In no
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/QueueBFSFundamentalCycleBasis.java]: typo: "If no such"

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/QueueBFSFundamentalCycleBasis.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ * vertices of the graph.
+ * </p>
+ * 
+ * @param <V> the vertex type
+ * @param <E> the edge type
+ *
+ * @author Dimitrios Michail
+ * @since October 2016
+ */
+public class QueueBFSFundamentalCycleBasis<V, E>
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/QueueBFSFundamentalCycleBasis.java]: Do we still need the class `PatonCycleBase`? This class seems very similar?

**Comment by jkinable** [comment]
> Note that I on purpose did not use a GraphPath here, as cycles are vectors in GF2 when being in the context of undirected cycle bases. This means that a cycle is just an edge set where each vertex has even degree

I agree that an undirected cycle can be walked in 2 different directions, and can thereby be represented by 2 equivalent directed cycles. Similarly, querying the edgeSource of an edge (u,v) in an undirected graph doesn't make too much sense. However, in the context of cycles it seems quite convenient to me that you can actually 'walk' the cycle. In your implementation, each cycle is represented as a `List<E>`. If the user wants to walk a cycle C represented by a `List<E> lc`, the user must first ensure that the edges in `lc` are ordered in a way that `edgeTarget(lc.get(i))==edgeSource(lc.get(i+1))`. After all, the `CycleBasisAlgorithm` interface doesn't guarantee that `lc` is ordered in a way that the cycle can be walked. A GraphPath on the other hand would offer this guarantee. Consequently, it seems more convenient to me that a cycle is represented as a GraphPath, even though I agree that technically a cycle is indeed just a vector in GF2. Returning a GraphPath instead of a List<E> would only require a minor modification to the method `buildFundamentalCycle(.)`.
Would you agree, or am I missing something?

**Comment by d-michail** [comment]
The problem here is that in the context of "cycle bases", a cycle is just a subset of the edges where every vertex has even degree. In disconnected graphs, such cycles cannot always be represented using a GraphPath.

Consider for example the following graph:

```
1-2   5-6
| |   | |
4-3   8-7
```

A cycle basis of this graph must have 2 cycles. One such basis is

```
cycle1: (1,2)(2,3)(3,4)(4,1)
cycle2: (1,2)(2,3)(3,4)(4,1)(5,6)(6,7)(7,8)(8,5)
```

if you add cycle1+cycle2 (in GF2, thus, symmetric difference) you get the cycle `(5,6)(6,7)(7,8)(8,5)`.

Cycle2 cannot be represented using a GraphPath.

The algorithms that we currently have do not return such cycles, but still the interface should be generic enough to return any cycle basis.

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/QueueBFSFundamentalCycleBasis.java (hunk 2 above)
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/QueueBFSFundamentalCycleBasis.java]: The version with the stack is more similar, but still class `PatonCycleBase` produces a different cycle basis. 

It does not work with multigraphs, but it produces a "weakly" fundamental cycle basis which in some cases might be shorter that the ones returned by the other algorithms. When this happens, it runs slightly faster (same complexity, less constants).

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AbstractFundamentalCycleBasis.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     * Given a non-tree edge and a spanning tree (forest) build a fundamental cycle.
+     * 
+     * @param graph the graph
+     * @param e a non-tree (forest) edge
+     * @param spanningForest the spanning tree (forest)
+     * @return a fundamental cycle
+     */
+    private Pair<List<E>, Double> buildFundamentalCycle(
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AbstractFundamentalCycleBasis.java]: nice idea!

**Comment by jkinable** [comment]
I agree that the general interface should be generic, and can therefore not return GraphPaths. Would it make sense to add a class level convenience method which returns GraphPaths? I had never seen or heard of a fundamental cycle space before you submitted this PR, so it's hard to judge for me what methods should be present in these classes. Wiki states: "In graph theory, a branch of mathematics, a cycle basis of an undirected graph is a set of _simple cycles_ that forms a basis of the cycle space of the graph.", hence my question. If not I'll merge this one.

**Comment by d-michail** [comment]
Indeed, when you restrict the definitions to graphs (due to [Veblen's Theorem](https://en.wikipedia.org/wiki/Veblen%27s_theorem)) a cycle basis is defined as a basis of the edge space using simple cycles (circuits). 

I usually keep the representation as an edge list for two main reasons: 
 1. it is more convenient in this case, as most of the time you perform set operations (disjoint union)
 2. in the case of directed graphs, an undirected cycle basis contains cycles which do not respect the edge directions.

Despite that I followed you suggestion and added an extra method to the interface which transforms the result to graph paths.

**Comment by jsichi** [comment]
Thanks!

**Pull request merged**